controllers/apim/virtualservice_controller.go

```diff
@@ -0,0 +1,216 @@
+
+	"github.com/go-logr/logr"
+	"github.com/kuadrant/kuadrant-controller/pkg/log"
+	"github.com/kuadrant/kuadrant-controller/pkg/reconcilers"
+)
+
+const (
+	KuadrantAuthProviderAnnotation = "kuadrant.io/auth-provider"
+)
+
+// VirtualServiceReconciler reconciles Istio's AuthorizationPolicy object
+type VirtualServiceReconciler struct {
+	*reconcilers.BaseReconciler
+	Scheme *runtime.Scheme
+}
+
+func (r *VirtualServiceReconciler) Reconcile(eventCtx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Logger().WithValues("VirtualService", req.NamespacedName)
+	ctx := logr.NewContext(eventCtx, logger)
+
+	virtualService := istionetworkingv1alpha3.VirtualService{}
+	if err := r.Client().Get(ctx, req.NamespacedName, &virtualService); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "failed to get VirtualService")
+		return ctrl.Result{}, err
+	}
+
+	if err := r.reconcileAuthPolicy(ctx, &virtualService); err != nil {
+		logger.Error(err, "failed to reconcile AuthorizationPolicy")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func getAuthPolicyName(vsName string) string {
+	return "source-virtualservice-" + vsName
+}
+
+func (r *VirtualServiceReconciler) reconcileAuthPolicy(ctx context.Context, vs *istionetworkingv1alpha3.VirtualService) error {
+	logger := r.Logger()
+	logger.Info("Reconciling AuthorizationPolicy")
+
+	authPolicy := istiosecurityv1beta1.AuthorizationPolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      getAuthPolicyName(vs.Name),
+			Namespace: vs.Namespace,
+		},
+	}
+
+	providerName, present := vs.GetAnnotations()[KuadrantAuthProviderAnnotation]
+	if !present {
+		return fmt.Errorf("Kuadrant auth-provider annotation not found")
+	}
+
+	// fill out the rules
+	authToRules := []*securityv1beta1.Rule_To{}
+	for _, httpRoute := range vs.Spec.Http {
+		for idx, matchRequest := range httpRoute.Match {
+			toRule := &securityv1beta1.Rule_To{
+				Operation: &securityv1beta1.Operation{},
+			}
+
+			toRule.Operation.Hosts = vs.Spec.Hosts
+			if normalizedURI := normalizeStringMatch(matchRequest.Uri); normalizedURI != "" {
+				toRule.Operation.Paths = append(toRule.Operation.Paths, normalizedURI)
+			}
+
+			if normalizedMethod := normalizeStringMatch(matchRequest.Method); normalizedMethod != "" {
+				// Looks like it's case-sensitive:
+				// https://istio.io/latest/docs/reference/config/security/normalization/#1-method-not-in-upper-case
+				method := strings.ToUpper(normalizedMethod)
+				toRule.Operation.Methods = append(toRule.Operation.Methods, method)
+			}
+
+			// If there is only regex stringmatches then we'll have bunch of repeated To rules with
+			// only same host filled into each. Following make sure only one field like that is present.
+			operation := toRule.Operation
+			if len(operation.Paths) == 0 && len(operation.Methods) == 0 && idx > 0 {
+				continue
+			}
+			authToRules = append(authToRules, toRule)
```

Comment:
this is controlling which paths and methods go to Authorino? Do we need that level or for now should we just send any request to the host?

Reply:
yes, specific paths. Better to specific I guess?

Reply:
This behavior is similar to AuthPolicy generated using OAS.